src/Confluent.Kafka/Deserializers.cs

```diff
@@ -69,7 +58,7 @@ public Null Deserialize(ReadOnlySpan<byte> data, bool isNull, SerializationConte
             {
                 if (!isNull)
                 {
-                    throw new DeserializationException("Deserializer<Null> may only be used to deserialize data that is null.");
+                    throw new ArgumentException("Deserializer<Null> may only be used to deserialize data that is null.");
```

Comment:
Since the deserializer is not being passed data from this applicatoin I would argue that ArgumentException is incorrect here, TypeException is better.

Reply:
This is an ArgumentException WRT this function, so it's what this function should throw. It's the job of additional layers to wrap and re-throw as appropriate and include this as an InnerException for deeper information.

Reply:
Yeah, but in practice this will be seen as an (inner) ArgumentException from Consume(), which  would typically indicate an argument error to Consume().
I think the Serdes-exceptions were good, very clear what was wrong.

Reply:
doesn't indicate that - the only time that would be the implication is if that exception is at the top level. Note that I think we may want to re-invent the Serialization and Deserialization exceptions in the future, but they should derive from ProduceException/ConsumeException. This is much better UX (as well as structurally better). Users who don't care can just do a blanket catch of ConsumeException, those who do can specialize. Specialization is possible now too of course, but the user needs to check the Error field (which is not idiomatic, but good enough for now and easily extended in this way). It's valuable for serde exceptions to derive from produce/consume exceptions because they carry with them context which may be useful in the handler, wherever that is.

Reply:
We could also have Serialization/Deserialization exceptions as now that don't derive from produce/consume exceptions, but they would still be thrown by the producer/consumer (with the inner exception as in this PR). but as noted, i think the additional context is useful, so we'll want to re-implement these classes to derive from produce/consume exceptions.